### PR TITLE
Make worker size and timeout configurable for execution pool

### PR DESCRIPTION
**Pull request by temaniarpit27**
# Description

Please provide a detailed description of what was done in this PR

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [ ] I have added sufficient documentation in code
- [ ] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: rpc/client.go [hunk 1]
```diff
@@ -112,7 +112,8 @@ func (c *Client) newClientConn(conn ServerCodec) *clientConn {
 	ctx := context.Background()
 	ctx = context.WithValue(ctx, clientContextKey{}, c)
 	ctx = context.WithValue(ctx, peerInfoContextKey{}, conn.peerInfo())
-	handler := newHandler(ctx, conn, c.idgen, c.services, NewExecutionPool(0))
+	// TODO - Check this
+	handler := newHandler(ctx, conn, c.idgen, c.services, NewExecutionPool(0), 10*time.Second)
```
Comment (JekaMas) [inline rpc/client.go:116]: why 10*time.Second? Maybe treat 0 as 'switched off'?

File: rpc/endpoints.go [hunk 2]
```diff
@@ -27,7 +28,7 @@ import (
 func StartIPCEndpoint(ipcEndpoint string, apis []API) (net.Listener, *Server, error) {
 	// Register all the APIs exposed by the services.
 	var (
-		handler    = NewServer()
+		handler    = NewServer(100, 10*time.Second)
```
Comment (JekaMas) [inline rpc/endpoints.go:31]: could we get rid of hard coded values?

File: rpc/handler.go [hunk 3]
```diff
@@ -64,28 +64,30 @@ type handler struct {
 	subLock    sync.Mutex
 	serverSubs map[ID]*Subscription
 
-	executionPool *SafePool
+	executionPoolRequestTimeout time.Duration
+	executionPool               *SafePool
```
Comment (JekaMas) [inline rpc/handler.go:68]: If we don't have request specific timeouts, then `executionPoolRequestTimeout` can be part of executionPool.

**Review by JekaMas — APPROVED** [review]

File: rpc/handler.go (hunk 3 above)
Comment (JekaMas) [reply rpc/handler.go:69]: also it should be atomic to make it changeable in runtime

File: internal/jsre/deps/web3.js [hunk 4]
```diff
@@ -1192,7 +1192,7 @@ module.exports = SolidityTypeInt;
     You should have received a copy of the GNU Lesser General Public License
     along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
 */
```
Comment (JekaMas) [inline internal/jsre/deps/web3.js:1194]: do we need this file changed?

File: node/api.go [hunk 5]
```diff
@@ -342,3 +343,57 @@ func (s *publicWeb3API) ClientVersion() string {
...
+	WSLimit   uint64
+}
+
+type ExecutionPoolRequestTimeout struct {
+	HttpLimit time.Duration
+	WSLimit   time.Duration
+}
+
+func (api *privateAdminAPI) GetExecutionPoolThreads() *ExecutionPoolThreads {
+	httpLimit := api.node.http.httpConfig.threads
```
Comment (JekaMas) [inline node/api.go:358]: if these values can be changed in runtime, then we should have either atomics or a mutex for them

File: node/api.go [hunk 6]
```diff
@@ -342,3 +343,57 @@ func (s *publicWeb3API) ClientVersion() string {
...
+	executionPoolThreads := &ExecutionPoolThreads{
+		HttpLimit: httpLimit,
+		WSLimit:   wsLimit,
+	}
+
+	return executionPoolThreads
+}
+
+func (api *privateAdminAPI) GetExecutionPoolRequestTimeout() *ExecutionPoolRequestTimeout {
+	httpLimit := api.node.http.httpConfig.requesttimeout
```
Comment (JekaMas) [inline node/api.go:370]: will `ChangeSize, ChangeTimeout` change this config values as well? If not we need either to start changing them or read actual values from SafePool.

File: rpc/server.go [hunk 7]
```diff
@@ -49,26 +50,34 @@ type Server struct {
...
 
 	// Register the default service providing meta information about the RPC service such
 	// as the services and methods it offers.
 	rpcService := &RPCService{server}
 	server.RegisterName(MetadataApi, rpcService)
-
 	return server
 }
 
+func (s *Server) SetExecutionPoolSize(n int) {
```
Comment (JekaMas) [inline rpc/server.go:73]: Could we also log that changes as in https://github.com/maticnetwork/bor/pull/712/files# (please ask @0xsharma for the latest changes)? It'd be great to have these logs similar.

File: internal/jsre/deps/web3.js (hunk 4 above)
Comment (temaniarpit27) [reply internal/jsre/deps/web3.js:1194]: no

File: node/api.go [hunk 8]
```diff
@@ -342,3 +343,61 @@ func (s *publicWeb3API) ClientVersion() string {
...
+		HttpLimit: httpLimit,
+		WSLimit:   wsLimit,
+	}
+
+	return executionPoolRequestTimeout
+}
+
+func (api *privateAdminAPI) SetWSExecutionPoolRequestTimeout(n int) *ExecutionPoolRequestTimeout {
+	api.node.ws.wsHandler.Load().(*rpcHandler).server.SetExecutionPoolRequestTimeout(time.Duration(n) * time.Second)
+	log.Info("modifying ws execution pool request timeout", "timeout", n)
```
Comment (JekaMas) [inline node/api.go:383]: let's make it warning level

File: rpc/endpoints.go [hunk 9]
```diff
@@ -27,7 +28,7 @@ import (
 func StartIPCEndpoint(ipcEndpoint string, apis []API) (net.Listener, *Server, error) {
 	// Register all the APIs exposed by the services.
 	var (
-		handler    = NewServer()
+		handler    = NewServer(10, time.Duration(10)*time.Second)
```
Comment (JekaMas) [inline rpc/endpoints.go:31]: do we need it hard-coded?

Comment (temaniarpit27) [reply rpc/endpoints.go:31]: no i am checking whether this is causing ipc issues

File: rpc/execution_pool.go [hunk 10]
```diff
@@ -8,27 +8,27 @@ import (
...
 
-const (
-	threads        = 100
-	requestTimeout = 0 //10 * time.Second
-)
-
 type SafePool struct {
 	executionPool *atomic.Pointer[workerpool.WorkerPool]
 	fastPath      bool
+	timeout       *atomic.Pointer[time.Duration]
```
Comment (JekaMas) [inline rpc/execution_pool.go:14]: if timeout was 0 then fastPath was true, if we then change timeout to non-zero value, fastPath incorrectly will be kept true.
I believe we need `fastPath` atomic and change `ChangeTimeout` `Change*`. Maybe have 1 RWMutex instead 2 atomic values?

Comment (temaniarpit27) [reply rpc/execution_pool.go:14]: fastpath is dependent upon size and not timeout I think

File: docs/cli/example_config.toml [hunk 11]
```diff
@@ -72,25 +72,30 @@ ethstats = ""                # Reporting URL of a ethstats service (nodename:sec
     api = ["eth", "net", "web3", "txpool", "bor"]  # API's offered over the HTTP-RPC interface
     vhosts = ["localhost"]                         # Comma separated list of virtual hostnames from which to accept requests (server enforced). Accepts '*' wildcard.
     corsdomain = ["localhost"]                     # Comma separated list of domains from which to accept cross origin requests (browser enforced)
+    ep-size = 100                                  # Maximum size of workers to run in rpc execution pool for HTTP requests (default: 100)
```
Comment (JekaMas) [inline docs/cli/example_config.toml:75]: let's make it 40

File: internal/jsre/deps/web3.js (hunk 4 above)
Comment (JekaMas) [reply internal/jsre/deps/web3.js:1194]: Could we revert formatting this file?

File: node/api.go [hunk 12]
```diff
@@ -342,3 +343,69 @@ func (s *publicWeb3API) ClientVersion() string {
...
+func (api *privateAdminAPI) SetWSExecutionPoolSize(n int) *ExecutionPoolSize {
+	api.node.ws.wsConfig.executionPoolSize = uint64(n)
+	api.node.ws.wsHandler.Load().(*rpcHandler).server.SetExecutionPoolSize(n)
+	log.Warn("Updating ws execution pool size", "threads", n)
+
+	return api.GetExecutionPoolSize()
+}
+
+func (api *privateAdminAPI) SetHttpExecutionPoolSize(n int) *ExecutionPoolSize {
+	api.node.http.httpConfig.executionPoolSize = uint64(n)
```
Comment (JekaMas) [inline node/api.go:406]: is it atomic? why we need both? direct `=` and `SetExecutionPoolSize` ?

**Review by JekaMas — APPROVED** [review]

File: node/api.go (hunk 12 above)
Comment (manav2401) [reply node/api.go:406]: One of them is updating the http config which does not need to be atomic. Another one is actually calling the RPC function to set the new size which will actually update the number of workers used by the execution pool. There's a mutex for that in the library.

Comment (JekaMas) [reply node/api.go:406]: could we have a few SetHttpExecutionPoolSize at the same time?

File: internal/jsre/deps/web3.js (hunk 4 above)
Comment (manav2401) [reply internal/jsre/deps/web3.js:1194]: Yes, will do

**Pull request closed without merging**